src/openms_gui/source/VISUAL/Spectrum2DCanvas.cpp

```diff
@@ -326,100 +326,85 @@ namespace OpenMS
-        cout << "BEGIN " << __PRETTY_FUNCTION__ << endl;
-        cout << "min spacing mz:" << min_spacing_mz << endl;
+      std::cerr << rt_min << ":" << rt_max <<"   " << mz_min << ":" << mz_max << "\n";
+      std::cerr << n_ms1_scans << "rt " << n_peaks_in_scan << "ms\n";
+      std::cerr << rt_pixel_count << " x " << mz_pixel_count << " px\n";
+      std::cerr << ratio_data2pixel_rt << " " << ratio_data2pixel_mz << " ratio\n";
 #endif
+      // minimum fraction of image expected to be filled with data
+      // if not reached, we upscale point size
+      const double MIN_COVERAGE = 0.2;
 
-        {
-          vector<float> rts;
-          for (Size i = 0; i != n_scans; ++i)
-          {
-            // skip non MS1 and empty spectra
-            Size ms_level = peak_map[i].getMSLevel();
-            Size n_peaks = peak_map[i].size();
-            if (ms_level != 1 || n_peaks == 0)
-            {
-              continue;
-            }
-            rts.push_back(peak_map[i].getRT());
-          }
-          sort(rts.begin(), rts.end());
-          if (rts.size() > 2)
-          {
-            average_spacing_rt = (rts[rts.size() - 1] - rts[0]) / (double)rts.size();
-          }
-        }
-      }
+      bool has_low_pixel_coverage = ratio_data2pixel_rt < MIN_COVERAGE || ratio_data2pixel_mz < MIN_COVERAGE;
 
-      // Determine whether several peaks are expected to be drawn on the same pixel
-      if (n_peaks_in_middle_scan > mz_pixel_count || n_ms1_scans > rt_pixel_count)
+      // Are several peaks expected to be drawn on the same pixel in either RT or m/z?
+      // --> thin out and show only maxima
+      // Also, we cannot upscale in this mode (since we operate on the buffer directly, i.e. '1 data point == 1 pixel'
+      if (!has_low_pixel_coverage & (n_peaks_in_scan > mz_pixel_count || n_ms1_scans > rt_pixel_count))
       {
-        // overlapping data points expected: draw maximum intensity
         paintMaximumIntensities_(layer_index, rt_pixel_count, mz_pixel_count, painter);
       }
       else
-      {
-        // calculate pixel width and height in rt/mz coordinates
-        QPoint p1, p2;
-        dataToWidget_(1, 1, p1);
-        dataToWidget_(0, 0, p2);
-        double pixel_width = abs(p1.x() - p2.x());
-        double pixel_height = abs(p1.y() - p2.y());
-
+      { // this is slower to paint, but allows scaling points
         // when data is zoomed in to single peaks these are visualized as circles
-        double pen_width = qMax( 1.0, isMzToXAxis()
-            ? min(pixel_width * min_spacing_mz, pixel_height * average_spacing_rt)
-            : min(pixel_width * average_spacing_rt, pixel_height * min_spacing_mz) );
 
-        // few data points expected: more expensive drawing of all datapoints (circles or points depending on zoom level)
+        const double MIN_PEN_SIZE = 1;  // minimum number of pixels for one data point
+        const double MAX_PEN_SIZE = 20; // maximum number of pixels for one data point
```

Comment:
same point as above, either we make it a parameter of the function or we put it into some namespace reserved for constants